torch/csrc/distributed/rpc/init.cpp

```diff
@@ -333,23 +346,71 @@ PyObject* rpc_init(PyObject* /* unused */) {
-                    .def(
-                        "wait",
-                        [&](FutureIValue& fut) {
-                          auto& pythonRpcHandler =
-                              PythonRpcHandler::getInstance();
-                          const auto& value = fut.wait();
-                          pybind11::gil_scoped_acquire ag;
-                          auto obj = torch::jit::toPyObject(value);
-                          pythonRpcHandler.handleException(obj);
-                          return obj;
-                        },
-                        py::call_guard<py::gil_scoped_release>(),
-                        R"(
+  shared_ptr_class_<FutureIValue>(module, "Future")
+      .def(
+          "wait",
+          [&](FutureIValue& fut) {
+            auto& pythonRpcHandler = PythonRpcHandler::getInstance();
+            const auto& value = fut.wait();
+            pybind11::gil_scoped_acquire ag;
+            auto obj = torch::jit::toPyObject(value);
+            pythonRpcHandler.handleException(obj);
+            return obj;
+          },
+          py::call_guard<py::gil_scoped_release>(),
+          R"(
 Wait on future to complete and return the object it completed with.
 If the future completes with an error, an exception is thrown.
-              )");
+          )")
+      .def(
+          "_then",
+          [&](FutureIValue& fut, py::function cb) {
+            // We need this an additional layer of wrapper here to guard the
+            // destruction of the py::function object. Because, the
+            // FutureIValue owns a reference to the py::function in its
+            // callback vector, but FutureIValue does not acquire GIL on
+            // destruction and it does not need to do so either. FutureIValue
+            // only acquires GIL if the IValue holds a py::object. However,
+            // nothing prevents applications to insert py::function as callbacks
+            // non-py::object FutureIValue. For example, the application can
+            // call a TorchScript function using rpc_async, and then append
+            // py::function as a callback. Hence, we use PythonFunction wrapper
+            // here to explicitly acquire GIL and decouple value destruction and
+            // callback destruction.
+            PythonFunction pf(std::move(cb));
+            return fut.then<IValue>([pf](const FutureIValue& fut) -> IValue {
+              if (fut.hasError()) {
+                throw std::runtime_error(c10::str(
+                    "Parent Future reported error: ",
+                    (*fut.error()).what()));
+              } else {
+                try {
+                  pybind11::gil_scoped_acquire ag;
+                  py::object ret =
+                      pf.func_(torch::jit::toPyObject(fut.constValue()));
+                  return jit::toIValue(std::move(ret), PyObjectType::get());
+                } catch (py::error_already_set& e) {
```

Comment:
Thanks for taking care of the error handling! This is great, though I was wondering what you think about making the python user's cb signature something like:
```
def callback(future_result, future_error):
    # User's responsibility to check and handle error
```

with this, we could avoid passing in the future here and instead pass in the result and optional error.

Of course, this is also fine, and we can revisit if user code needs custom error handling.

Reply:
> For example, there may be cases where the user does not want to throw if the parent future failed.

I see, this makes sense. In this case, should we let the Python callback API also takes the original future instead of the return value? Otherwise, the two args will be mutual exclusive optional return value and optional error?

Reply:
This would be good too, although it can be done as a follow up too since I don't feel there's a need to block this diff on that. The API is private for now as well.